Selector generation for batch/v1 Job

erictune opened the pull request.

erictune committed:
Added Selector Generation to Job.

Added selector generation to Job's
strategy.Validate, right before validation.
Can't do in defaulting since UID is not known.

Added a validation to Job to ensure that the generated
labels and selector are correct when generation was requested.
This happens right after generation, but validation is in a better
place to return an error.

Adds "manualSelector" field to batch/v1 Job to control selector generation.
Adds same field to extensions/__internal.  Conversion between those two
is automatic.

Adds "autoSelector" field to extensions/v1beta1 Job.  Used for storing batch/v1 Jobs
    - Default for v1 is to do generation.
    - Default for v1beta1 is to not do it.
    - In both cases, unset == false == do the default thing.

Release notes:
Added batch/v1 group, which contains just Job, and which is the next
version of extensions/v1beta1 Job.

The changes from the previous version are:
- Users no longer need to ensure labels on their pod template are unique to the enclosing
  job (but may add labels as needed for categorization).
- In v1beta1, job.spec.selector was defaulted from pod labels, with the user responsible for uniqueness.
  In v1, a unique label is generated and added to the pod template, and used as the selector (other
  labels added by user stay on pod template, but need not be used by selector).
- a new field called "manualSelector" field exists to control whether the new behavior is used,
  versus a more error-prone but more flexible "manual" (not generated) seletor.  Most users
  will not need to use this field and should leave it unset.

Users who are creating extensions.Job go objects and then posting them using the go client
will see a change in the default behavior.  They need to either stop providing a selector (relying on
selector generation) or else specify "spec.manualSelector" until they are ready to do the former.

erictune committed:
Regenerate

erictune committed:
Conversion between manualSelector and autoSelector

erictune commented:
@lavalamp I am seeing a failure in `TestRoundTripTypes` in `make test WHAT=pkg/api` which I don't understand and which doesn't seem related to my changes.  Could someone from @kubernetes/goog-csi  pull this down and run `make test WHAT=pkg/api`.

erictune commented:
Also, while I am still going to add a few more tests, I think now would be a good time to do your first review of this code.

erictune commented:
@soltysh your feedback wanted as well.
@sdminonne you may also be interested.

When this merges, and you rebase the ScheduledJob controller (or Workflow controller) onto it, it won't work without some changes to your job creation logic.  You'll need to remove the selector from the POST'ed job, so that the selector generation logic can generate one for you.

erictune commented:
This **has** to go into 1.2 or else we need to revert the change to move job to v1.

lavalamp commented on pkg/apis/batch/v1/types.go:
```diff
@@ -70,9 +71,22 @@ type JobSpec struct {
 	Selector *LabelSelector `json:"selector,omitempty"`
 
+	// ManualSelector controls generation of pod labels and pod selectors.
+	// Leave `noAuthSelector` unset unless you are certain what you are doing.
```
noAuthSelector?

erictune replied:
fixing

lavalamp commented:
As for the test failure, you need to add to the custom fuzz function for job. https://github.com/kubernetes/kubernetes/blob/17325ef6ef61918d49a28d615c717bccec5b5a71/pkg/api/testing/fuzzer.go#L160

sdminonne commented:
@erictune, note taken. Thanks

soltysh commented:
@erictune will look into it first thing in the morning

erictune commented:
Okay, this is ready for review.

lavalamp commented:
Can you do a rebase? I think you picked up a couple extra commits.

erictune commented:
done.

On Tue, Feb 23, 2016 at 3:43 PM, Daniel Smith <notifications@github.com>
wrote:

> Can you do a rebase? I think you picked up a couple extra commits.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/21776#issuecomment-187969725>
> .
>

lavalamp commented on docs/user-guide/jobs.md:
```diff
@@ -323,6 +308,70 @@ Here, `W` is the number of work items.
+You cannot update the job because these fields are not updatable.
+Therefore, you delete job `old` but leave its pods
+running, using `kubectl delete jobs/old-one --cascase=false`.
+Before deleting it, you make a note of what selector it uses:
```
Shouldn't you first create the new job? Otherwise, a completion could happen while no job is running?

erictune replied:
it is level-triggered, not edge-triggered.

lavalamp commented on examples/job/expansions/README.md:
```diff
@@ -75,12 +72,14 @@ In a real use case, the processing would be some substantial computation, such a
-key `item`, distinguishes individual jobs in the group.  Each Job object needs to have
-a unique label that no other job has.  This is it.
-Neither of these label keys are special to kubernetes -- you can pick your own label scheme.
+This Job and it s Pod template have a label: `jobgroup=jobexample`.  There is nothing special
```
s/it s/its/

lavalamp commented on pkg/apis/extensions/v1beta1/conversion.go:
```diff
@@ -277,3 +279,103 @@ func Convert_v1beta1_ReplicaSetSpec_To_extensions_ReplicaSetSpec(in *ReplicaSetS
+	// autoSelector has opposite meaning as manualSelector.
+	// in both cases, unset means false, and unset is always preferred to false.
+	// unset vs set-false distinction is not preserved.
+	manualSelector := bool(in.ManualSelector != nil && *in.ManualSelector)
```
The bool cast shouldn't be necessary.

lavalamp commented on pkg/apis/extensions/v1beta1/conversion.go:
```diff
@@ -277,3 +279,103 @@ func Convert_v1beta1_ReplicaSetSpec_To_extensions_ReplicaSetSpec(in *ReplicaSetS
+	if autoSelector {
+		out.AutoSelector = new(bool)
+		*out.AutoSelector = true
+	}
```
`else {
  out.AutoSelector = nil
}`

lavalamp commented on pkg/apis/extensions/v1beta1/conversion.go:
```diff
@@ -277,3 +279,103 @@ func Convert_v1beta1_ReplicaSetSpec_To_extensions_ReplicaSetSpec(in *ReplicaSetS
+	if manualSelector {
+		out.ManualSelector = new(bool)
+		*out.ManualSelector = true
+	}
```
Same comments

lavalamp replied:
It's traditional to write a test for custom conversion functions.

ixdy commented:
@k8s-bot unit test this again please, issue #IGNORE (I accidentally broke one of the Jenkins slaves)

lavalamp commented:
Mostly minor comments.

soltysh commented:
Mostly nits, but otherwise LGTM.

erictune commented:
Addressed lavalamp and soltysh comments.

lavalamp commented:
LGTM

erictune commented:
@k8s-bot please e2e again flake #21751

erictune committed:
Update docs and examples to batch/v1 Job

Documented manualSelector field.

Documented that you do not need to provide a selector
or unique labels with batch/v1 Job.

Updated all Job examples to apiVersion: batch/v1

Updated all Job examples to use generated selectors.

soltysh commented:
@k8s-bot unit test this github flake: #21975

soltysh commented:
LGTM

erictune commented:
@k8s-bot please e2e test again issue #21975

erictune commented:
@k8s-bot please e2e test again issue #14968

erictune commented:
Thanks @fabioy, that was a long submit queue!

fabioy commented:
No prob. It's hard to get these big changes in sometimes...

liggitt commented:
From the commit message:
> Users who are creating extensions.Job go objects and then posting them using the go client will see a change in the default behavior.  They need to either stop providing a selector (relying on selector generation) or else specify "spec.manualSelector" until they are ready to do the former.

I am creating an object like this:
```go
job := &extensions.Job{}
job.Name = "testjob"
job.Spec.Template.Labels = map[string]string{"foo": "bar"}
job.Spec.Template.Spec = ...
job.Spec.Template.Spec.RestartPolicy = kapi.RestartPolicyNever
_, err = client.Extensions().Jobs(testutil.Namespace()).Create(job)
```
and am getting this error:
> Job.extensions "testjob" is invalid: spec.selector: Invalid value: {"matchLabels":{"controller-uid":"70badb08-e8ea-11e5-8dc1-acbc32c1ca87","foo":"bar"}}: `selector` not auto-generated

Should that be working?
